zip-0302.rst

```diff
@@ -0,0 +1,81 @@
+::
+
+  ZIP: 302
+  Title: Standardized Memo Field Format
+  Owner: Jack Grigg <jack@electriccoin.co>
+  Original-Author: Jay Graber
+  Status: Draft
+  Category: Standards
+  Created: 2017-02-08
+  License: MIT
+
+
+Abstract
+========
+
+This ZIP describes a proposed specification for a standardized format for displaying content within the encrypted memo field of shielded transactions.
+
+Motivation
+==========
+
+A well-defined standard for formatting content within the encrypted memo field would help expand the use cases by providing a structure for different types of data. Users and third-party services could benefit from a standardized formatting convention that defines the type and length of the data contained within.
+
+A proposed common use case could be a standard encoding for sender’s preferred return address.
+
+Specification
+===============
+
+Section 5.5 of the Zcash protocol specification [#protocol-notept]_ defines three cases
+for the encoding of a memo field:
+
+* a UTF-8 human-readable string [#Unicode]_, padded by appending zero bytes; or
+* the byte ``0xF6`` followed by 511 ``0x00`` bytes, indicating "no memo"; or
+* any other sequence of 512 bytes starting with a byte value ``0xF5`` or greater
+  (which is therefore not a valid UTF-8 string), as specified in ZIP 302.
+
+This ZIP refines the specification of the third case.
+
++ If the first byte (byte 0)'s value is 0xF4 or smaller, then the reader MUST:
+     + strip any trailing zero bytes
+     + decode it as UTF-8 (replacing any incorrect UTF-8-encoded byte sequences with the replacement character U+FFFD), and display it to the user as a human-readable string.
++ If byte 0's value is 0xF5, then the reader MUST:
+     + Interpret the next few bytes (1 to 9 of them) as a 64-bit unsigned variable-length integer [#Bitcoin-CompactSize]_, and use it as an arbitrary application-defined "type" field.
+     + Interpret the next bytes (1 to 2 of them) as a 16-bit unsigned ULEB, and use it as the length field. (The length can be at most 510 bytes due to the overall memo length, and that is why the length field can only be 1 or 2 bytes.)
+     + If 1 + the number bytes used for the type field + the number of bytes used for the length field + the length > 512 then error out, i.e. do not do any further processing of the memo, and do not return any information about the memo to the caller other than the fact that it was incorrectly formatted.
```

Comment:
Would it reduce complexity or simplify anything to strictly require that the length field always be 2 bytes? Because, in practice, whenever we use 1 byte the memo, by definition, it doesn't need the extra space. Conversely, when the memo is potentially maxing out the available bytes, we need two bytes ANYWAY to represent its length. In other words, leveraging a ULEB here doesn't seem to provide any savings.

Reply:
I'm not sure that backwards compatibility with those implementations is necessarily something that justifies the complexity of CompactSize.

Reply:
I prefer leaving this as it is; a CompactSize is not significantly more complicated.